process_hierarchy.py

```diff
@@ -0,0 +1,478 @@
+                parent_id \
+                ) values ( \
+                %(id)s, \
+                now(), \
+                %(type)s, \
+                %(name)s, \
+                %(status)s, \
+                %(display_name)s, \
+                %(parent_id)s \
+                )",
+                {
+                    "id": hierarchy_id,
+                    "type": rec["type"],
+                    "name": rec["name"],
+                    "display_name": rec["display_name"],
+                    "status": rec["status"],
+                    "parent_id": rec["parent_id"],
+                },
+            )
+            # update the dictionary
+            current_dict[hierarchy_id]["type"] = rec["type"]
+            current_dict[hierarchy_id]["name"] = rec["name"]
+            current_dict[hierarchy_id]["display_name"] = rec["display_name"]
+            current_dict[hierarchy_id]["status"] = rec["status"]
+            current_dict[hierarchy_id]["parent_id"] = rec["parent_id"]
+            current_dict[hierarchy_id]["still_active"] = True
+    else:
+        # create a new record with a new id
+        cursor.execute(
+            "insert into hierarchy_db.hierarchy_rec ( \
+            id, \
+            create_ts, \
+            type, \
+            name, \
+            status, \
+            display_name, \
+            parent_id \
+            ) values ( \
+            nextval(hierarchy_db_id_seq), \
+            now(), \
+            %(type)s, \
+            %(name)s, \
+            %(status)s, \
+            %(display_name)s, \
+            %(parent_id)s \
+            )",
+            {
+                "type": rec["type"],
+                "name": rec["name"],
+                "display_name": rec["display_name"],
+                "status": rec["status"],
+                "parent_id": rec["parent_id"],
+            },
+        )
+        # lastrowid doesn't seem to work with nextval()
+        # id = cursor.lastrowid
+        result = (
+            moc_db_helper_functions.exec_fetchall(
+                cursor,
+                "select id from hierarchy_rec hr where hr.type=%(type)s and hr.name=%(name)s",
+                {"type": rec["type"], "name": rec["name"]},
+            )
+        )[0]
+        hierarchy_id = None
+        if result and "id" in result:
+            hierarchy_id = result["id"]
+        if hierarchy_id:
+            current_dict[hierarchy_id] = copy.copy(rec)
+            current_dict[hierarchy_id]["id"] = id
+            current_dict[hierarchy_id]["still_active"] = True
+    pprint.pprint(current_dict)
+
+
+def create_hierarchy_dictionary(cursor, sql_stmt, params):
+    """
+    This creates a dictionary of the records returned
+
+    currently the primary key is (id, create_ts) in order to
+    track history -
+    """
+    records = moc_db_helper_functions.exec_fetchall(cursor, sql_stmt, params)
+    ret_dict = {}
+    for rec in records:
+        ret_dict[rec["id"]] = copy.deepcopy(rec)
```

Comment:
You don't need `copy.deepcopy` here; you can just set:

```
ret_dict[rec["id"]] = rec
```

You're not trying to preserve the original value of `rec` for any reason.

You could write this as a dictionary comprehension:

```
return {rec["id"]: rec for rec in records}
```